File: 프로그래머스/1/42862. 체육복/체육복.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <cstring>

using namespace std;

int visited[34];

int solution(int n, vector<int> lost, vector<int> reserve) {
    int answer = 0;
    for(int i=1;i<=n;i++) {
        visited[i]=1;
    }
    
    for(int i=0;i<lost.size();i++) {
        visited[lost[i]]-=1;
    }
    for(int i=0;i<reserve.size();i++) {
        visited[reserve[i]]+=1;
    }
    
    for(int i=1;i<=n;i++){
        if(visited[i]==0) {
            
            if(visited[i-1]==2) {
                visited[i-1]=1;
                visited[i]=1;
            } else if (visited[i+1]==2) {
                visited[i]=1;
                visited[i+1]=1;
            }
        }
    }
    
    for(int i=1;i<=n;i++) {
        if(visited[i]) {
            answer+=1;
        }
    }
    return answer;
}

File: 프로그래머스/1/42862. 체육복/체육복.java
import java.util.*;

class Solution {
    
    int visited[] = new int[34];
    int lost_visited[] = new int[34];
    
    public int solution(int n, int[] lost, int[] reserve) {
        int answer = 0;
        
        for(int i=0;i<lost.length;i++) {
            lost_visited[lost[i]]=1;
        }
        
        for(int i=1;i<=n;i++) {
            if(lost_visited[i]==1) continue;
            visited[i]=1;
        }
        
        for(int i=0;i<reserve.length;i++) {
            if(lost_visited[reserve[i]]==1) {
                lost_visited[reserve[i]]=0;
                visited[reserve[i]]=2;
            }
        }
        Arrays.sort(reserve);
        
        for(int i=0;i<reserve.length;i++) {
            
            if(visited[reserve[i]]==2) continue;
            
            if(lost_visited[reserve[i]]==1) {
                lost_visited[reserve[i]]=0;
                visited[reserve[i]]=1;
                continue;
            }
            
            if(reserve[i]==1) {
                if(lost_visited[reserve[i]+1]==1) {
                    lost_visited[reserve[i]+1]=0;
                    visited[reserve[i]+1]=1;
                }
            } else if(reserve[i]==n) {
                if(lost_visited[reserve[i]-1]==1) {
                    lost_visited[reserve[i]-1]=0;
                    visited[reserve[i]-1]=1;
                }
            } else {
                if(lost_visited[reserve[i]-1]==1) {
                    lost_visited[reserve[i]-1]=0;
                    visited[reserve[i]-1]=1;
                } else {
                    if(lost_visited[reserve[i]+1]==1) {
                        lost_visited[reserve[i]+1]=0;
                        visited[reserve[i]+1]=1;
                    }
                }
            }
        }
        
        for(int i=1;i<=n;i++) {
            if(visited[i]==1 || visited[i]==2) {
                answer+=1;
            }
        }
        return answer;
    }
}